File: apps/pos_app/lib/features/inventory/weighted_average_cost.dart
import 'package:decimal/decimal.dart';

/// Weighted-average unit cost (VND integers). Mirrors API helper.
int weightedAverageCost({
  required Decimal oldQty,
  required int oldAvgVnd,
  required Decimal receiptQty,
  required int unitCostVnd,
}) {
  if (receiptQty <= Decimal.zero) {
    return oldAvgVnd < 0 ? 0 : oldAvgVnd;
  }
  if (oldQty <= Decimal.zero) {
    return unitCostVnd < 0 ? 0 : unitCostVnd;
  }
  final oldQ = oldQty.toDouble();
  final recQ = receiptQty.toDouble();
  final avg = (oldQ * oldAvgVnd + recQ * unitCostVnd) / (oldQ + recQ);
  final rounded = avg.round();
  return rounded < 0 ? 0 : rounded;
}

File: apps/api/src/inventory/weighted-average-cost.ts
/** Weighted-average unit cost (VND integers). */
export function weightedAverageCost(
  oldQty: number,
  oldAvgVnd: number,
  receiptQty: number,
  unitCostVnd: number,
): number {
  if (!(receiptQty > 0) || !Number.isFinite(receiptQty)) {
    return Math.max(0, Math.round(oldAvgVnd));
  }
  if (!(oldQty > 0)) {
    return Math.max(0, Math.round(unitCostVnd));
  }
  const totalQty = oldQty + receiptQty;
  const avg = (oldQty * oldAvgVnd + receiptQty * unitCostVnd) / totalQty;
  return Math.max(0, Math.round(avg));
}
